api/gateway-operator/v2alpha1/controlplane_types.go

```diff
@@ -0,0 +1,264 @@
+// that the ControlPlane is responsible for configuring.
+type ControlPlaneDataPlaneTargetType string
+
+const (
+	// ControlPlaneDataPlaneTargetURL indicates that the DataPlane target is a URL.
+	ControlPlaneDataPlaneTargetURL ControlPlaneDataPlaneTargetType = "url"
+
+	// ControlPlaneDataPlaneTargetName indicates that the DataPlane target is a name
+	// of a DataPlane resource managed by the operator.
+	ControlPlaneDataPlaneTargetName ControlPlaneDataPlaneTargetType = "name"
+)
+
+// ControlPlaneAdminAPI defines the configuration for the DataPlane Kong Admin API.
+type ControlPlaneAdminAPI struct {
+	// Workspace indicates the Kong Workspace to use for the ControlPlane.
+	// If left empty then no Kong workspace will be used.
+	//
+	// +optional
+	Workspace string `json:"workspace,omitempty"`
+}
+
+// ControlPlaneController defines a controller state for the ControlPlane.
+// It overrides the default behavior as defined in the deployed operator version.
+//
+// +apireference:kgo:include
+type ControlPlaneController struct {
+	// Name is the name of the controller.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name,omitempty"`
+
+	// Enabled indicates whether the controller is enabled or not.
+	//
+	// +kubebuilder:validation:Required
+	Enabled *bool `json:"enabled,omitempty"`
+}
+
+// ControlPlaneFeatureGate defines a feature gate state for the ControlPlane.
+// It overrides the default behavior as defined in the deployed operator version.
+//
+// +apireference:kgo:include
+type ControlPlaneFeatureGate struct {
+	// Name is the name of the feature gate.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name,omitempty"`
+
+	// Enabled indicates whether the feature gate is enabled or not.
+	//
+	// +kubebuilder:validation:Required
+	Enabled *bool `json:"enabled,omitempty"`
+}
+
+// ControlPlaneStatus defines the observed state of ControlPlane
+//
+// +apireference:kgo:include
+type ControlPlaneStatus struct {
+	// Conditions describe the current conditions of the Gateway.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Scheduled", status: "Unknown", reason:"NotReconciled", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// FeatureGates is a list of effective feature gates for this ControlPlane.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=32
+	FeatureGates []ControlPlaneFeatureGate `json:"featureGates,omitempty"`
+
+	// Controllers is a list of enabled and disabled controllers for this ControlPlane.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=32
+	Controllers []ControlPlaneController `json:"controllers,omitempty"`
```

Comment:
Since the status reflects the current status of the CP, what do you think about setting here just the enabled controllers? I'm just thinking of a list of plain strings with the controller names.

Reply:
On one hand that makes sense but OTOH having also the disabled controllers (or feature gates) makes it very easy for users to see which items they can enable in the spec. Without this information they have to consult the documentation.

WDYT?

Reply:
Makes sense. So, your opinion is to list here ALL the available controllers along with their enablement?

Reply:
All, as defined in particular operator version. So whatever the operator is aware of. This will allow the user to know just by looking at the status what is available for enabling/disabling.

I'll then go ahead and make the change.